### ISSUE TITLE
lint suggestion: unused_macro_fragment

### ISSUE THREAD

**Issue by lolbinarycat** [rust-lang/rust-clippy#14863]
### What it does

if a macro fragment (the things in macro rules starting with `$`) that does not start with `_` is not used, generate a warning.

### Advantage

same advantages as linting on unused function arguments.

when copy/paste refactoring code into a function, it is possible to forget to change a hardcoded value to use a function argument, and the same mistake can also happen when refactoring code into a macro.

### Drawbacks

maybe this should be in rustc instead of clippy, so it can be in the `unused` group.

### Example

```rust
macro_rules! foo {
    ($op:tt) => {};
}

foo!(bar);
```

Could be written as:

```rust
macro_rules! foo {
    ($_op:tt) => {};
}

foo!(bar);
```

**Comment by Jarcho** [comment]
This feels like it should be in rustc. I'm actually not sure why it doesn't do this already.

**Comment by nyurik** [comment]
There was a request for this in rustc many years ago - https://github.com/rust-lang/rust/issues/18820 - but it was closed. I suspect this could be too disruptive for the core rustc community?

Note that some may call this an unused macro parameter / variable rather than a fragment.  The unused branch of a macro might already exist as a lint.
